referenceframe/frame_system.go

```diff
@@ -308,6 +312,67 @@ func (sfs *simpleFrameSystem) FrameSystemSubset(newRoot Frame) (FrameSystem, err
 	return newFS, nil
 }
 
+// FrameSystemToPCD takes in a framesystem and returns a map where all elements are
+// the point representation of their geometry type with respect to the world.
+func FrameSystemToPCD(system FrameSystem, inputs map[string][]Input) (map[string][]r3.Vector, error) {
+	vectorMap := make(map[string][]r3.Vector)
+	geoMap, err := FrameSystemGeometries(system, inputs)
+	if err != nil {
+		return nil, err
+	}
+	for name, geosInFrame := range geoMap {
+		geos := geosInFrame.geometries
+		aggregatePoints := []r3.Vector{}
+		for _, g := range geos {
+			asPoints := g.ToPoints(defaultPointDensity)
+			aggregatePoints = append(aggregatePoints, asPoints...)
+		}
+		vectorMap[name] = aggregatePoints
+	}
+	return vectorMap, nil
+}
+
+// FrameSystemGeometries takes in a framesystem and returns a map where all elements
+// are GeometriesInFrame modified to be with respect to the world.
+func FrameSystemGeometries(system FrameSystem, inputs map[string][]Input) (map[string]*GeometriesInFrame, error) {
+	geoMap := make(map[string]*GeometriesInFrame)
+	for _, name := range system.FrameNames() {
+		currentFrame := system.Frame(name)
+		parent, err := system.Parent(currentFrame)
+		if err != nil {
+			return nil, err
+		}
+		currentInput := inputs[name]
+		// if currentInput is nil and DoF != 0 we chose to omit the
+		// frame entirely as this would return the frame's geometries
+		// in their home or "zero" position, and not in their
+		// current position.
+		if currentInput == nil && len(currentFrame.DoF()) == 0 {
+			currentInput = []Input{}
+		}
+		if currentInput != nil {
+			geosInFrame, err := currentFrame.Geometries(currentInput)
+			if err != nil {
+				return nil, err
+			}
+			if geosInFrame.geometries != nil {
```

Comment:
We should not be checking if the private struct member is nil here. `geometries` could be nil, or it could be a non-nil zero-length map, the latter of which wouldn't be caught here.

Instead, checking `if len(geosInFrame.Geometries) > 0 {` will do the right thing